internal/storage/cluster/cluster.go

```diff
@@ -0,0 +1,286 @@
+}
+
+func (cs *ClusterStorage) clearClaimsIDCookie(r *http.Request, w http.ResponseWriter) {
+	http.SetCookie(w, &http.Cookie{
+		Name:     storage.ClaimsIdCookie,
+		Value:    "",
+		Path:     "/",
+		Domain:   cs.authenticator.GetCookieDomain(r),
+		Expires:  time.Now().Local().Add(time.Hour * -1),
+		Secure:   true,
+		HttpOnly: true,
+	})
+}
+
+func (cs *ClusterStorage) generateClaimsID() string {
+	id := make([]byte, storage.ClaimsIdLength)
+	_, _ = rand.Read(id)
+	return fmt.Sprintf("%x", id)
+}
+
+func (cs *ClusterStorage) generateHMAC(claimsId string) string {
+	h := hmac.New(sha256.New, cs.HmacSecret)
+	h.Write([]byte(claimsId))
+	return fmt.Sprintf("%x", h.Sum(nil))
+}
+
+func (cs *ClusterStorage) checkHMAC(claimsId string, mac string) bool {
+	return hmac.Equal([]byte(mac), []byte(cs.generateHMAC(claimsId)))
+}
+
+func (cs *ClusterStorage) getAndValidateClaimsId(r *http.Request) (string, error) {
+	c, err := r.Cookie(storage.ClaimsIdCookie)
+	if err != nil {
+		return "", err
+	}
+	sp := strings.Split(c.Value, ":")
+	if len(sp) != 2 {
+		return "", CookieValidationError("cookie is malformed")
+	}
+	claimsId := sp[0]
+	mac := sp[1]
+
+	if !cs.checkHMAC(claimsId, mac) {
+		return "", CookieValidationError("cookie failed authentication")
+	}
+
+	return claimsId, nil
+}
+
+func (cs *ClusterStorage) getUserInfoFromSecret(s *corev1.Secret) (*v1alpha1.UserInfo, error) {
+	data, ok := s.Data[storage.UserInfoKey]
+	if !ok {
+		return nil, SecretError("userinfo data missing from secret")
+	}
+
+	userinfo := &v1alpha1.UserInfo{}
+	if err := json.Unmarshal(data, userinfo); err != nil {
+		return nil, fmt.Errorf("%v: %w", SecretError("error parsing userinfo"), err)
+	}
+
+	return userinfo, nil
+}
+
+func (cs *ClusterStorage) storeUserInfo(claimId string, info *v1alpha1.UserInfo) error {
+	data, err := json.Marshal(info)
+	if err != nil {
+		return fmt.Errorf("error mashelling userinfo: %w", err)
+	}
+	s := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: storage.SecretNameBase,
+			Labels: map[string]string{
+				storage.ClaimsLabel:   "true",
+				storage.ClaimsIDLabel: claimId,
+			},
+		},
+		Data: map[string][]byte{
+			storage.UserInfoKey: data,
+		},
+	}
+
+	if _, err := cs.Client.CoreV1().Secrets(cs.Namespace).Create(s); err != nil {
+		return fmt.Errorf("%v: %w", SecretError("error creating secret"), err)
+	}
+	return nil
+}
+
+func (cs *ClusterStorage) getSecrets() (*corev1.SecretList, error) {
+	return cs.Client.CoreV1().Secrets(cs.Namespace).List(metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=true", storage.ClaimsLabel),
+	})
+}
+
+func (cs *ClusterStorage) getSecretByClaim(claimsId string) (*corev1.Secret, error) {
+	secrets, err := cs.getSecrets()
+	if err != nil {
+		return nil, fmt.Errorf("error getting secret list: %w", err)
+	}
+
+	for _, s := range secrets.Items {
+		cid, ok := s.ObjectMeta.Labels[storage.ClaimsIDLabel]
+		if !ok {
+			logger.Errorf("found managed secret not containing claimID")
```

Comment:
should the user take action on this error, and is there enough info to do so?

Reply:
adding the secret namespace/name to the logging output.